### DISCUSSION TITLE
sideset enable bit in PIO

### DISCUSSION THREAD

**Discussion by webdeb** [micropython/micropython#13743] [RP2040 / Pico]
The datasheet has this paragraph about the sideset enable bit. I am struggeling to find how to do it micropython.

```
Whether to use the most significant of these bits as an enable.
Side-set takes place on instructions where the enable is high.
If there is no enable bit, every instruction on that state machine will perform a side-set, if SIDESET_COUNT is nonzero. 
This is configured by EXECCTRL_SIDE_EN.
```

I've two machines toggling the same pin. And this is fine, while one machine is waiting for another pin, the other should be able to toggle the sideset pin. Right now I've only accomplished it using the `set` instruction. But this takes the limited instructions space.

Maybe someone has an idea how to accomplish this. Thank you.

**Comment by rkompass** [comment]
In Micropython to use all available bits for side set you have to put a `.side()` on all instructions. By that the  EXECCTRL_SIDE_EN will be automatically configured to so that no enable bit be used. And as a result you have the remaining bits available for delays. Otherwise the enable bit is used from the same set of 5 (iirc) bits and you only have the number of remaining bits -1 available for delay.
If you have to side set only one pin you might want to use delays up to 15. If delays up to 7 (i.e. 3 bits) are sufficient then you do not have to consider the above. If you want to use delays up to 15 (i.e. 4 bits) then put a `.side()` on every instruction.

In that case the other machine toggling the pin would not work - apparently, as both machines would do the side-set.
The datasheet states:

> For each individual GPIO output (level and direction separately), PIO considers all 8 writes that may have occurred on
> that cycle, and applies the write from the highest-numbered state machine. If the same state machine performs a SET
> /OUT and a side-set on the same GPIO simultaneously, the side-set is used. If no state machine writes to this GPIO
> output, its value does not change from the previous cycle.

So the highe- numbered machine could still side-set, while the lower-numbered would wait (doing the side-set) only in vain.

Again If you do not have to use delays > 7 this is not relevant and you may use side-set (and add the `.side()` to the instruction only where it is really needed. MP would automatically configure EXECCTRL_SIDE_EN for using enable bit then.

**Reply by webdeb** [reply]
Thats the point. I would like to use `side set` like `set`, which means, set the pin only on this exact instruction, and on the next cpu cycle another machine regardless of sm id should be able to rewrite it. As I understand, this can be accomplished by setting `EXECCTRL_SIDE_EN` to 0.

**Reply by webdeb** [reply]
> MP would automatically configure EXECCTRL_SIDE_EN for using enable bit then.

seems to not configure it automatically

**Reply by rkompass** [reply]
What I described happens [here](https://github.com/micropython/micropython/blob/master/ports/rp2/modules/rp2.py)
where in line 69 ff:
```py
self.sideset_opt = self.num_sideset != self.num_instr
if self.sideset_opt:
    self.prog[_PROG_EXECCTRL] |= 1 << 30
    self.sideset_count += 1
```
the comparison is done if every instruction has a `.side()` in it and if not so Bit 30 of _PROG_EXECCTRL is set.
This is SIDE_EN: From Datasheet:
> If 1, the MSB of the Delay/Side-set instruction field is used as side-set enable, rather than a side-set data bit. This
> allows instructions to perform side-set optionally, rather than on every instruction, but the maximum possible sideset width is reduced from 5 to 4. Note that the value of PINCTRL_SIDESET_COUNT is inclusive of this enable bit.

> As I understand, this can be accomplished by setting EXECCTRL_SIDE_EN to 0.

> seems to not configure it automatically

I see, here seems to be the misunderstanding:  EXECCTRL_SIDE_EN has to be 1 to achieve the desired behaviour.
